src/cascadia/TerminalApp/AppLogic.cpp

```diff
@@ -400,6 +400,36 @@ namespace winrt::TerminalApp::implementation
         ShowDialog(dialog);
     }
 
+    // Method Description:
+    // - Displays a dialog if the "Touch, Keyboard and Handwriting Panel
+    //   Service" is disabled. If it is, we want to warn the user, because they
+    //   won't be able to type in the Terminal.
+    // - Only one dialog can be visible at a time. If another dialog is visible
+    //   when this is called, nothing happens. See ShowDialog for details
+    void AppLogic::_ShowKeyboardServiceDisabledDialog()
+    {
+        auto title = RS_(L"KeyboardServiceWarningTitle");
+        auto text = RS_(L"KeyboardServiceWarningText");
+        auto buttonText = RS_(L"Ok");
+
+        Controls::TextBlock warningsTextBlock;
+        // Make sure you can copy-paste
+        warningsTextBlock.IsTextSelectionEnabled(true);
+        // Make sure the lines of text wrap
+        warningsTextBlock.TextWrapping(TextWrapping::Wrap);
+
+        warningsTextBlock.Inlines().Append(_BuildErrorRun(text, Application::Current().as<::winrt::TerminalApp::App>().Resources()));
+        warningsTextBlock.Inlines().Append(Documents::LineBreak{});
+

```

Comment:
is there any way to not have another copy of our dialog code lol

Reply:
Like, maybe we can have this one just be in the `.xaml` file directly? And bind its values via UID?

Reply:
Lower in this thread we mention using GetServiceName to get the actual OS-localized name of the service, which might actually prevent us from binding the resources with Uid's. Since that seems like a preferable fix, I'm gonna keep this one _not in XAML_.

Reply:
:| it should be possible to bind the text run to a function call like we do for versions

Reply:
meh fine, I'll do it

Reply:
Thanks! When i think about the things that are hard in code (using a themeresource! (Or any resource actually!) Constructing styled text runs!) it makes me wish we had _more_ markup. It seems worse to dig deeper in a place we know is bad. :)